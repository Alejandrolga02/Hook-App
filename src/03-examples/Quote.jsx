import PropTypes from "prop-types";
import { useLayoutEffect, useRef, useState } from "react";

export const Quote = ({ quote, author }) => {
	const paragraphRef = useRef();
	const [boxSize, setBoxSize] = useState({ width: 0, height: 0 });

	useLayoutEffect(() => {
		const { height, width } = paragraphRef.current.getBoundingClientRect();
		setBoxSize({ height, width });
	}, []);

	return (
		<>
			<blockquote
				className="blockquote text-end"
				style={{
					display: "flex",
					flexDirection: "column",
					justifyContent: "end",
				}}>
				<p className="mb-3" ref={paragraphRef}>
					{quote}
				</p>
				<footer className="blockquote-footer">{author}</footer>
			</blockquote>

			<code>{JSON.stringify(boxSize)}</code>
		</>
	);
};

Quote.propTypes = {
	quote: PropTypes.string.isRequired,
	author: PropTypes.string.isRequired,
};
