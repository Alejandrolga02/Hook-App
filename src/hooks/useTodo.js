import { useCallback, useEffect, useReducer, useState } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const init = () => {
	return JSON.parse(localStorage.getItem('todos')) || [];
}

export const useTodo = () => {
	const [todos, dispatch] = useReducer(todoReducer, [], init);
	const [pendingTasks, setPendingTasks] = useState(0);

	const addTodo = useCallback(
		(description) => {
			dispatch({
				type: '[TODO] Add Item',
				payload: {
					id: new Date().getTime(),
					done: false,
					description,
				}
			});
		}
	, []);

	const deleteItem = useCallback(
		(id) => {
			return dispatch({
				payload: id,
				type: '[TODO] Delete Item',
			})
		}
	, []);

	const toggleItem = useCallback(
		(id) => {
			return dispatch({
				payload: id,
				type: '[TODO] Toggle Item',
			})
		}
	, []);

	useEffect(() => {
		localStorage.setItem('todos', JSON.stringify(todos));

		const pendingTasksNumber = todos.reduce((counter, task) => {
			if (!task.done) {
				return counter + 1;
			}
			return counter;
		}, 0);
	
		setPendingTasks(pendingTasksNumber);
	}, [todos]);

	return {
		todos,
		addTodo,
		deleteItem,
		toggleItem,
		pendingTasks
	}
}