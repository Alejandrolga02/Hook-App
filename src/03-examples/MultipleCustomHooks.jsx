import { useCounter, useFetch } from "../hooks";
import { LoadingQuote, Quote } from "./";

export const MultipleCustomHooks = () => {
	const { counter, increment } = useCounter(1);
	const { data, isLoading, hasError } = useFetch(`https://api.breakingbadquotes.xyz/v1/quotes/${counter}`);
	const { quote, author } = !!data && data[0];

	if (hasError) {
		return <div className="alert alert-danger text-center">Something ocurred contact the admin</div>;
	}

	return (
		<>
			<h1>Breaking Bad Quotes</h1>
			<hr />

			{isLoading ? <LoadingQuote /> : <Quote quote={quote} author={author} />}

			<button onClick={() => increment(1)} className="btn btn-dark" disabled={isLoading}>
				Next Quote
			</button>
		</>
	);
};
