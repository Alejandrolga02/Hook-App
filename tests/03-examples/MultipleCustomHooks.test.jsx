import { fireEvent, render, screen } from "@testing-library/react";
import { MultipleCustomHooks } from "../../src/03-examples";
import { useFetch } from "../../src/hooks/useFetch";
import { useCounter } from "../../src/hooks/useCounter";

jest.mock("../../src/hooks/useFetch");
jest.mock("../../src/hooks/useCounter");

describe("tests in <MultipleCustomHooks />", () => {
	const mockIncrement = jest.fn();
	useCounter.mockReturnValue({
		counter: 1,
		increment: mockIncrement
	});

	beforeEach(() => {
		jest.clearAllMocks();
	})

	test("should show default component", () => {
		useFetch.mockReturnValue({
			data: null,
			isLoading: true,
			hasError: null,
		});

		render(<MultipleCustomHooks />);
		expect(screen.getByText("Loading..."));
		expect(screen.getByText("Breaking Bad Quotes"));

		const nextButton = screen.getByRole("button", { name: "Next Quote" });
		expect(nextButton.disabled).toBeTruthy();
	});

	test("should show a Quote", () => {
		useFetch.mockReturnValue({
			data: [{ author: "Alejandro", quote: "Test" }],
			isLoading: false,
			hasError: null,
		});

		const { container } = render(<MultipleCustomHooks />);
		expect(container).toMatchSnapshot();
		expect(screen.getByText("Test")).toBeTruthy();
		expect(screen.getByText("Alejandro")).toBeTruthy();

		const nextButton = screen.getByRole("button", { name: "Next Quote" });
		expect(nextButton.disabled).toBeFalsy();
		screen.debug();
	});

	test("should call increment function", () => {
		useFetch.mockReturnValue({
			data: [{ author: "Alejandro", quote: "Test" }],
			isLoading: false,
			hasError: null,
		});

		render(<MultipleCustomHooks />);

		const nextButton = screen.getByRole("button", { name: "Next Quote" });
		fireEvent.click(nextButton);

		expect(mockIncrement).toHaveBeenCalled();
	});
});
