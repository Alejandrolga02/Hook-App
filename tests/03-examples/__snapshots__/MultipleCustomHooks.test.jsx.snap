// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests in <MultipleCustomHooks /> should show a Quote 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <blockquote
    class="blockquote text-end"
    style="display: flex; flex-direction: column; justify-content: end;"
  >
    <p
      class="mb-3"
    >
      Test
    </p>
    <footer
      class="blockquote-footer"
    >
      Alejandro
    </footer>
  </blockquote>
  <code>
    {"height":0,"width":0}
  </code>
  <button
    class="btn btn-dark"
  >
    Next Quote
  </button>
</div>
`;
